src/vehicle_move.cpp

```diff
@@ -116,6 +117,261 @@ int vehicle::slowdown( int at_velocity ) const
+    bool opt_safe_vel = safe_ground_velocity( true );
+    float cur_load_approx = static_cast<float>( std::min( accel_demand,
+                            opt_accel ) )  / std::max( opt_accel, 1 );
+    float cur_load_alternator = std::min( 0.01f, static_cast<float>( alternator_load ) / 1000 );
+
+    cur_state.engines_hash = 31 * c_engines.size();
+
+    for( size_t i = 0; i < c_engines.size(); ++i ) {
+        cur_state.engines_hash += i * c_engines[i] * 13;
+
+        if( is_engine_on( c_engines[i] ) ) {
+            prev_mask |= 1 << i;
+            bool is_electric = is_engine_type( c_engines[i], fuel_type_battery );
+            int fu = engine_fuel_usage( c_engines[i] ) * ( cur_load_approx + ( is_electric ? 0 :
+                     cur_load_alternator ) );
+            opt_fuel_usage += fu;
+            if( is_electric ) {
+                opt_net_echarge_rate -= fu;
+            }
+        }
+    }
+    cur_state.created = calendar::turn;
+    cur_state.engines_mask = prev_mask;
+    cur_state.battery_percent = battery_level_percent;
+    cur_state.battery_net_charge_rate = opt_net_echarge_rate;
+    cur_state.velocity = avg_velocity;
+    cur_state.load = cur_load_approx + cur_load_alternator;
+    if( smart_controller_state ) {
+        cur_state.gas_engine_last_turned_on = smart_controller_state->gas_engine_last_turned_on;
+    }
+    cur_state.gas_engine_shutdown_forbidden = gas_engine_shutdown_forbidden;
+
+    int opt_mask = prev_mask; // save current engine state, because it will be temporarily modified
+
+    // if vehicle state has not change, skip actual optimization
+    if( smart_controller_state &&
+        std::abs( smart_controller_state->velocity - cur_state.velocity ) < 100 &&
+        std::abs( smart_controller_state->battery_percent - cur_state.battery_percent ) <= 2 &&
+        std::abs( smart_controller_state->load - cur_state.load ) < 0.1 && // load diff < 10%
+        smart_controller_state->engines_hash == cur_state.engines_hash &&
+        smart_controller_state->battery_net_charge_rate == cur_state.battery_net_charge_rate &&
+        smart_controller_state->engines_mask == cur_state.engines_mask &&
+        // reevaluate cache if when cache was created, gas engine shutdown was forbidden, but now it's not
+        !( smart_controller_state->gas_engine_shutdown_forbidden && !gas_engine_shutdown_forbidden )
+      ) {
+        smart_controller_state->created = calendar::turn;
+        return;
+    }
+
+    // trying all combinations of engine state (max 31 iterations for 5 engines)
+    for( int mask = 1; mask < 1 << c_engines.size(); ++mask ) {
+        if( mask == prev_mask ) {
+            continue;
+        }
+
+        bool gas_engine_to_shut_down = false;
+        for( size_t i = 0; i < c_engines.size(); ++i ) {
+            bool old_state = static_cast<bool>( ( prev_mask & ( 1 << i ) ) != 0 );
+            bool new_state = static_cast<bool>( ( mask & ( 1 << i ) ) != 0 );
```

Comment:
```suggestion
            const bool new_state = ( mask & ( 1 << i ) ) != 0;
```
Same on the line above.